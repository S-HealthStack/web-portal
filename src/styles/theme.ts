import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
