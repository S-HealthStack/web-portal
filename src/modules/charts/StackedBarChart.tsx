import React, { useRef, useState, useMemo, useCallback } from 'react';
import * as d3 from 'd3';
import _sumBy from 'lodash/sumBy';
import _slice from 'lodash/slice';
import styled from 'styled-components';

import { typography, px, colors, theme } from 'src/styles';
import Tooltip from 'src/common/components/Tooltip';
import { XAxis, YAxis, Area } from './common-components';
import { TooltipProps, NoResponsesLabel, NO_RESPONSES_LABEL } from './common-helpers';

const TooltipContent = styled.div`
  display: flex;
  flex-direction: column;
`;

const TooltipRowContent = styled.div`
  display: flex;
`;

const TooltipNameBlock = styled.div`
  display: flex;
  margin-right: ${px(8)};
  ${typography.labelSemibold};
  color: ${colors.surface};
`;

const TooltipDataBlock = styled.div`
  display: flex;
  flex-direction: column;
  ${typography.labelRegular};
  color: ${colors.surface};
`;

interface DataItem {
  percentage: number;
  scaleValue: number;
}

export type BarChartProps = {
  data: DataItem[];
  width: number;
  height: number;
  color?: string;
  minScale?: number;
  maxScale?: number;
  className?: string;
};

const DEFAULT_COLOR = '#00B0D7';

const TOOLTIP_MARGIN_TOP = 37;

const StackedBarChart = ({
  width,
  height,
  data,
  color = DEFAULT_COLOR,
  minScale = 1,
  maxScale = 10,
  className,
}: BarChartProps) => {
  const svgRef = useRef<SVGSVGElement>(null);

  const [tooltipProps, setTooltipProps] = useState<TooltipProps>(null);

  const x = useMemo(
    () => d3.map(data, (d, index) => _sumBy(_slice(data, 0, index + 1), 'percentage')),
    [data]
  );
  const y = useMemo(() => [0, 1], []);

  const xRange = useMemo(() => [0, width], [width]);
  const yRange = useMemo(() => [height, 0], [height]);

  const xScale = useMemo(() => d3.scaleLinear([0, 100], xRange).nice(), [xRange]);
  const yScale = useMemo(() => d3.scaleLinear(y, yRange).nice(), [y, yRange]);

  const getAreaKeyByIndex = (index: number) => `area-${index}`;

  const onAreaMouseEnter = useCallback(
    (event: React.MouseEvent<SVGPathElement, MouseEvent>, index: number) => {
      if (!svgRef.current) {
        return;
      }

      const svgRect = svgRef.current.getBoundingClientRect();
      const areaId = getAreaKeyByIndex(index);
      const areaRect = (
        d3.select(svgRef.current).select(`#${areaId}`).node() as Element
      )?.getBoundingClientRect();

      const tooltipContent = (
        <TooltipContent>
          <TooltipRowContent>
            <TooltipNameBlock>SCALE VALUE:</TooltipNameBlock>
            <TooltipDataBlock>{data[index].scaleValue}</TooltipDataBlock>
          </TooltipRowContent>
          <TooltipRowContent>
            <TooltipNameBlock>PERCENTAGE:</TooltipNameBlock>
            <TooltipDataBlock>{`${Math.round(data[index].percentage)}%`}</TooltipDataBlock>
          </TooltipRowContent>
        </TooltipContent>
      );
      const position =
        (svgRect && (areaRect.left + 200 > svgRect.left + svgRect.width ? 'l' : 'r')) || 'l';
      const pointX = areaRect.left - svgRect.left - 10 + areaRect.width / 2;

      setTooltipProps({
        content: tooltipContent,
        point: [pointX, areaRect.top - svgRect.top + TOOLTIP_MARGIN_TOP],
        position,
      });
    },
    [data]
  );

  const onAreaMouseLeave = useCallback(() => {
    setTooltipProps(null);
  }, []);

  const isEmptyState = useMemo(() => data.every((d) => d.percentage === 0), [data]);

  return (
    <div className={className}>
      <svg ref={svgRef} width={width} height={height}>
        <XAxis key="x-axis" xScale={xScale} removeDomain yOffset={height} ticks={0} tickSize={0} />
        <YAxis key="y-axis" yScale={yScale} xOffset={0} removeDomain ticks={0} tickSize={0} />
        {data.map((d, index) => (
          <Area
            key={getAreaKeyByIndex(index)}
            id={getAreaKeyByIndex(index)}
            data={[
              { x: x[index - 1] || 0, y0: 0, y1: 1 },
              { x: x[index], y0: 0, y1: 1 },
            ]}
            xScale={xScale}
            yScale={yScale}
            color={color}
            fillOpacity={(1 / (maxScale - minScale + 1)) * (index + 1)}
            visible
            onMouseEnter={(event) => onAreaMouseEnter(event, index)}
            onMouseLeave={onAreaMouseLeave}
          />
        ))}
        {isEmptyState && (
          <rect x="0px" y="0px" width={width} height={height} fill={theme.colors.background} />
        )}
        {isEmptyState && (
          <NoResponsesLabel
            x="50%"
            y="50%"
            textAnchor="middle"
            dominantBaseline="middle"
            width={width}
            height={height}
          >
            {NO_RESPONSES_LABEL}
          </NoResponsesLabel>
        )}
      </svg>
      <Tooltip
        static
        content={tooltipProps?.content}
        point={tooltipProps?.point}
        show={!!tooltipProps}
        position={tooltipProps?.position}
        arrow
      />
    </div>
  );
};

export default styled(StackedBarChart)``;
