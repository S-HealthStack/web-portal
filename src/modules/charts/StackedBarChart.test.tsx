import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { act, findByTestId, render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components/';
import { userEvent } from '@storybook/testing-library';
import theme from 'src/styles/theme';
import StackedBarChart from 'src/modules/charts/StackedBarChart';

describe('StackedBarChart', () => {
  it('should render', async () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <StackedBarChart
          width={1000}
          height={1000}
          data={[
            {
              percentage: 90,
              scaleValue: 0,
            },
            {
              percentage: 10,
              scaleValue: 1,
            },
          ]}
        />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const area = await findByTestId(baseElement, 'area-0');
    expect(area).toBeInTheDocument();

    act(() => {
      userEvent.hover(area);
    });

    act(() => {
      userEvent.unhover(area);
    });
  });

  it('[NEGATIVE] should render with empty data', () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <StackedBarChart width={1000} height={1000} data={[]} />
      </ThemeProvider>
    );

    expect(baseElement).toBeDefined();
  });
});
