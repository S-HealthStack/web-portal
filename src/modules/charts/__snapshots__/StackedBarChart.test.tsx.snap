// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackedBarChart should render 1`] = `
.c2 {
  display: inline-block;
  position: relative;
}

.c0 > .tick > text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
}

.c1 > .tick > text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
}

<body>
  <div>
    <div
      class=""
    >
      <svg
        height="1000"
        width="1000"
      >
        <g
          class="c0 xAxis undefined"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="middle"
          transform="translate(0, 1000)"
        />
        <g
          class="c1 yAxis"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="start"
          transform="translate(0,0)"
        />
        <path
          class="area"
          d="M0,0L900,0L900,1000L0,1000Z"
          data-testid="area-0"
          fill="#00B0D7"
          fill-opacity="0.1"
          id="area-0"
        />
        <path
          class="area"
          d="M900,0L1000,0L1000,1000L900,1000Z"
          data-testid="area-1"
          fill="#00B0D7"
          fill-opacity="0.2"
          id="area-1"
        />
        <rect
          fill-opacity="0"
          id="area-label-0"
          x="450px"
          y="37px"
        />
        <rect
          fill-opacity="0"
          id="area-label-1"
          x="950px"
          y="37px"
        />
      </svg>
      <span
        class="c2"
        data-testid="tooltip-container"
      />
    </div>
  </div>
</body>
`;
